### ISSUE TITLE
Rule Proposal: no-proto-method-return

### ISSUE THREAD

**Issue by ljosberinn** [sindresorhus/eslint-plugin-unicorn#1098]
It's easy to forget actually calling a prototype method which leads to possibly hard-to-debug errors.

Since e.g. many array methods accept a callback or only an optional parameter, this could be handled with an allowlist per global object, e.g. `Array.prototype.map` wouldnt need to be checked for.

## Fail

```js
function foo(myString) {
  return string.slice(3).toLowerCase
}

function foo(myArr) {
  return myArr.map(val => val ** 2).reverse
}
```

## Pass

```js
function foo(myString) {
  return string.slice(3).toLowerCase()
}

function foo(myArr) {
  return myArr.map(val => val ** 2).reverse()
}
```

**Comment by sindresorhus** [comment]
> Since e.g. many array methods accept a callback or only an optional parameter, this could be handled with an allowlist per global object, e.g. Array.prototype.map wouldnt need to be checked for.

I'm not sure I understand what you're saying here.

**Comment by ljosberinn** [comment]
Sorry, what I was trying to say is that its imo more likely to accidentally do
```js
function foo(myArr) {
  return myArr.reverse
}
```
than on a `map`/`reduce`/`filter` because those require a callback, while `reverse` just misses `()`. 
Because of that, the rule might not have to check whether `map`/`reduce`/`filter` are used properly, but only those that are easier to trip over.

**Comment by sindresorhus** [comment]
Can you make a list of methods that are easy to miss that we should include in this rule?

**Comment by ljosberinn** [comment]
easy to miss - require no arguments passed:
- entries
- pop
- reverse
- shift
- toString

grey area - have an optional first argument and can be missed:
- concat
- join
- of
- slice
- sort
- toLocaleString

hard to miss - require at least one argument:
- copyWithin
- every
- filter
- find
- findIndex
- flatMap
- forEach
- from
- includes
- indexOf
- isArray
- lastIndexOf
- map
- push
- reduce
- reduceRight
- some
- splice

**Comment by fisker** [comment]
I think this should not only work for `return`, any of these methods not called are invalid.

## Fail

```js
while (array.pop) {}
```

```js
if (array.find) {}
````

---

Even this

```js
const push = foo.push;
push.call(bar, 1, 2);
```

Because

```js
const push = Array.prototype.push;
push.call(bar, 1, 2);
```

is better.

**Comment by fisker** [comment]
`no-uncalled-method`?

**Comment by sindresorhus** [comment]
> no-uncalled-method?

👍

**Comment by sindresorhus** [comment]
This is now accepted.
